### PR TITLE
[13.0][MIG] product_catalog_report

### PR DESCRIPTION
**Pull request by MiquelRForgeFlow**
Migrating a prehistoric module. From api v5 to api v13 in one fell swoop.

### CODE REVIEW

File: product_catalog_report/wizard/wizard_product_catalog.py [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+    _name = "res.partner.product_catalog"
+    _description = "Wizard Product Catalog"
+
+    def _get_language(self):
+        lang_obj = self.env["res.lang"]
+        languages = lang_obj.search([("active", "=", True)])
+        return [(lang.code, lang.name) for lang in languages]
+
+    def _get_default_report_lang(self):
+        partner_obj = self.env["res.partner"]
```
Comment (kos94ok-3D) [inline product_catalog_report/wizard/wizard_product_catalog.py:18]: изображение

File: product_catalog_report/report/product_catalog.py [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+
+    def _get_imagepath(self, product_id):
+        attach_ids = (
+            self.env["ir.attachment"]
+            .search(
+                [("res_model", "=", "product.product"), ("res_id", "=", product_id)],
+            )
+            .ids
+        )
+        datas = self.env["ir.attachment"].read(attach_ids)
```
Comment (kos94ok-3D) [inline product_catalog_report/report/product_catalog.py]: Why not use `search_read` with `limit=1`

File: product_catalog_report/report/product_catalog.py [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+    def _getProducts(self, category, lang):
+        prod_tmpIDs = (
+            self.env["product.template"].search([("categ_id", "=", category)]).ids
+        )
+        prod_ids = (
+            self.env["product.product"]
+            .search([("product_tmpl_id", "in", prod_tmpIDs)])
+            .ids
+        )
+        return self.env["product.product"].with_context(lang=lang).read(prod_ids)
```
Comment (kos94ok-3D) [inline product_catalog_report/report/product_catalog.py]: `search_read`

File: product_catalog_report/report/product_catalog.py [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+    def _get_currency(self):
+        return self.env["res.users"].browse([self.uid])[0].company_id.currency_id.name
+
+    def _get_packaging_title(self, product_id, index):
+        packaging_ids = (
+            self.env["product.packaging"]
+            .search([("product_id", "=", product_id)], limit=4)
+            .ids
+        )
+        packs = self.env["product.packaging"].read(packaging_ids, ["name"])
```
Comment (kos94ok-3D) [inline product_catalog_report/report/product_catalog.py]: The same

File: product_catalog_report/report/product_catalog.py [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+                return s
+        return " "
+
+    def _get_packaging_value(self, product_id, index):
+        packaging_ids = (
+            self.env["product.packaging"]
+            .search([("product_id", "=", product_id)], limit=4)
+            .ids
+        )
+        packs = self.env["product.packaging"].read(packaging_ids, ["qty"])
```
Comment (kos94ok-3D) [inline product_catalog_report/report/product_catalog.py]: The same

**Review by kos94ok-3D — CHANGES_REQUESTED** [review]

File: product_catalog_report/wizard/wizard_product_catalog.py (hunk 1 above)
Comment (MiquelRForgeFlow) [reply product_catalog_report/wizard/wizard_product_catalog.py:18]: That's an Odoo convention, not an OCA convention.

File: product_catalog_report/report/product_catalog.py [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+
+    def setCat(self, cats):
+        lst = []
+        for cat in cats:
+            if cat not in lst:
+                lst.append(cat)
+                category = self.env["product.category"].read([cat])
+                if category[0]["child_id"]:
+                    lst.extend(self.setCat(category[0]["child_id"]))
+        return lst
```
Comment (kos94ok-3D) [inline product_catalog_report/report/product_catalog.py]: Maybe better rewrite like:
<pre>
def setCat(self, cat_ids):
    categories = self.env['product.category'].browse(cat_ids)
    return (categories + categories.child_id).ids
</pre>

File: product_catalog_report/report/product_catalog.py [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+        price = pricelist.with_context(uom=False).price_get(product_id, 1.0, None)[
+            pricelist_id
+        ]
+        if not price:
+            price = 0.0
+        return price
+
+    def _get_desc(self, template_id):
+        if template_id:
+            prodtmpl = self.env["product.template"].read([template_id])[0]
```
Comment (kos94ok-3D) [inline product_catalog_report/report/product_catalog.py]: prodtmpl = self.env["product.template"].browse(template_id).read()[0]

**Review by kos94ok-3D — CHANGES_REQUESTED** [review]

File: product_catalog_report/views/product_report.xml [hunk 8]
```diff
@@ -0,0 +1,11 @@
+<?xml version="1.0" encoding="utf-8" ?>
+<odoo>
+    <report
+        id="report_product_catalog"
+        string="Product Catalog"
+        model="res.partner"
+        name="product_catalog_report.report_product_catalog"
+        file="product_catalog_report.report_product_catalog"
+        report_type="qweb-pdf"
```
Comment (kos94ok-3D) [inline product_catalog_report/views/product_report.xml:9]: add `menu="False"`

File: product_catalog_report/wizard/product_wizard.xml [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+        </field>
+    </record>
+    <record id="action_res_partner_product_catalog" model="ir.actions.act_window">
+        <field name="name">Partner Product Catalog</field>
+        <field name="res_model">res.partner.product_catalog</field>
+        <field name="type">ir.actions.act_window</field>
+        <field name="view_mode">form</field>
+        <field name="view_id" ref="res_partner_product_catalog_view" />
+        <field name="context">{'record_id': active_id}</field>
+        <field name="target">new</field>
```
Comment (kos94ok-3D) [inline product_catalog_report/wizard/product_wizard.xml:37]: add ` <field name="binding_model_id" ref="base.model_res_partner"/>`

**Review by kos94ok-3D — COMMENTED** [review]
If you want keep old behave (Allow run report only from wizard form)

File: product_catalog_report/report/product_catalog.xml [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+                                >
+                                    <tr>
+                                        <td colspan="2">
+                                            <span t-esc="image_url(p['id'])" />
+                                        </td>
+                                        <td colspan="2" />
+                                    </tr>
+                                    <tr>
+                                        <td colspan="2">
+                                            <span t-esc="p['code'] - p['name']" />
```
Comment (kos94ok-3D) [inline product_catalog_report/report/product_catalog.xml]: изображение
So
`<span t-esc="p['code']"/> - <span t-esc="p['name']"/>`

**Comment by MiquelRForgeFlow** [comment]
Many thanks @kos94ok-3D . Now it works fine locally. Although I don't know why that error in travis ci is showing, because it seems unrelated.

I suppose that what's maybe left to do is to apply cosmetic changes in the report template to improve the visualization. I am open to suggestions.

**Comment by kos94ok-3D** [comment]
sorry, but I don't know how it should look

File: product_catalog_report/views/product_report.xml [hunk 11]
```diff
@@ -0,0 +1,12 @@
+<?xml version="1.0" encoding="utf-8" ?>
+<odoo>
+    <report
+        id="report_product_catalog"
```
Comment (kos94ok-3D) [inline product_catalog_report/views/product_report.xml]: rename to action_report_product_catalog

File: product_catalog_report/wizard/wizard_product_catalog.py [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+    )
+
+    def create_product_catalog(self):
+        self.ensure_one()
+        data = {}
+        data["ids"] = self.env.context.get("active_ids", [])
+        data["model"] = "res.partner"
+        data["form"] = self.read(["report_lang", "categories"])[0]
+        return self.env.ref(
+            "product_catalog_report.report_product_catalog"
```
Comment (kos94ok-3D) [inline product_catalog_report/wizard/wizard_product_catalog.py]: rename to product_catalog_report.action_report_product_catalog

**Review by kos94ok-3D — COMMENTED** [review]
I don`t sure but maybe the problem in xml_id

**Comment by MiquelRForgeFlow** [comment]
Thanks, now it's green :green_apple:

**Pull request closed without merging**